File: kombi-lib/src/main/java/com/sgnatiuk/cartesian/CartesianProduct.java
package com.sgnatiuk.cartesian;

import com.sgnatiuk.Splittable;

import java.math.BigInteger;
import java.util.stream.Stream;

public interface CartesianProduct<T> extends Iterable<T>, Splittable<CartesianProduct<T>> {
    BigInteger combinationsCount();
    Stream<T> stream();
}

File: kombi-lib/src/main/kotlin/com/sgnatiuk/cartesian/CartesianProduct.kt
package com.sgnatiuk.cartesian

import com.sgnatiuk.Splittable
import java.math.BigInteger
import java.util.stream.Stream

interface CartesianProduct<T> : Iterable<T>, Splittable<CartesianProduct<T>> {
    val combinationsCount: BigInteger

    fun stream(): Stream<T>
}
